### PR TITLE
Eliza and heimdal tracking + improvements

### PR DESCRIPTION
**Pull request by camhashemi**
@leogrim @joshm1 @ryanpbrewster

### CODE REVIEW

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/MessagingCommander.scala [hunk 1]
```diff
@@ -679,17 +679,10 @@ class MessagingCommander @Inject() (
 
     val moreContext = new HeimdalContextBuilder()
     val orgParticipantsFuture = Future.sequence(orgIds.map { oid =>
-      shoebox.hasOrganizationMembership(oid, userId).flatMap {
```
Comment (camhashemi) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/MessagingCommander.scala:682]: we check whether the user has `GROUP_MESSAGING` permission for each org, this shouldn't be necessary

File: kifi-backend/modules/heimdal/app/com/keepit/heimdal/EventContextHelper.scala [hunk 2]
```diff
@@ -50,12 +52,13 @@ class EventContextHelperImpl @Inject() (
 
   def getOrgUserValues(orgId: Id[Organization]): Future[Seq[(String, ContextData)]] = {
     val shoeboxValuesFut = getOrgTrackingValues(orgId)
-    val membersFut = shoebox.getOrganizationMembers(orgId)
 
-    val userWithMostClickedKeepsFut = membersFut.map(helprankCommander.getUserWithMostClickedKeeps)
+    val userWithMostClickedKeepsFut = orgMemberWithMostClickedKeepsCache.getOrElseFuture(OrgMemberWithMostClickedKeepsKey(orgId)) {
+      val membersFut = shoebox.getOrganizationMembers(orgId)
+      membersFut.map(helprankCommander.getUserWithMostClickedKeeps)
```
Comment (camhashemi) [inline kifi-backend/modules/heimdal/app/com/keepit/heimdal/EventContextHelper.scala:59]: this query was timing out in heimdal, seeing if caching will help relieve stress on him

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/tracking/EventProxyController.scala [hunk 3]
```diff
@@ -18,13 +18,14 @@ class EventProxyController @Inject() (
...
   def track() = MaybeUserAction(parse.tolerantJson) { request =>
+    import com.keepit.common.core._
     SafeFuture("event proxy") {
       val sentAt = clock.now()
       request.body.as[Seq[JsObject]].foreach { rawEvent =>
         val rawEventType = (rawEvent \ "event").as[String]
         val (eventType, intendedEventOpt) = getEventType(rawEventType)
         val builder = heimdalContextBuilderFactoryBean.withRequestInfo(request)
-        (rawEvent \ "properties").asOpt[HeimdalContext] match {
+        (rawEvent \ "properties").as[JsObject].nonNullFields.asOpt[HeimdalContext] match {
```
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/tracking/EventProxyController.scala:30]: clients were sending null fields, so `HeimdalContext.reads` was breaking. unfortunately this happened silently due to the lack of airbrakes / exceptions below. I'm actually going to add an airbrake there, unless there are objections.

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/MessagingCommander.scala [hunk 4]
```diff
@@ -679,17 +679,10 @@ class MessagingCommander @Inject() (
...
-          moreContext += ("messagedWholeOrgId", ContextStringData(oid.toString))
-          shoebox.getOrganizationMembers(oid)
-        case false =>
-          Future.successful(Set.empty[Id[User]])
-      }
+      shoebox.getOrganizationMembers(oid)
     }).map(_.flatten)
 
-    moreContext.data.get("messagedWholeOrgId").collect { case orgId: ContextStringData => log.info(s"[OrgMessageTracking] should be tracking messagedWholeOrgId=$orgId") }
+    if (orgIds.nonEmpty) moreContext += ("messagedAllOrgId", orgIds.last.toString)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/MessagingCommander.scala]: Why the last org id?

File: kifi-backend/modules/eliza/app/com/keepit/eliza/controllers/internal/ElizaController.scala [hunk 5]
```diff
@@ -137,4 +137,10 @@ class ElizaController @Inject() (
     Ok(Json.toJson(totalMessages))
   }
 
+  def getParticipantsByThreadExtId(threadId: String) = Action { request =>
```
Comment (ryanpbrewster) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/controllers/internal/ElizaController.scala]: Why a `String`? I think that `ExternalId[_]` is bindable.

File: kifi-backend/modules/heimdal/app/com/keepit/heimdal/EventContextHelper.scala [hunk 6]
```diff
@@ -108,6 +111,19 @@ class EventContextHelperImpl @Inject() (
...
     }
   }
+
+  def getOrganizationIdByExtThreadId(threadExtId: String): Future[Option[Id[Organization]]] = {
+    eliza.getParticipantsByThreadExtId(threadExtId).flatMap { userIds =>
+      shoebox.getOrganizationsForUsers(userIds).map { orgsByUserId =>
+        orgsByUserId.values.reduceLeftOption[Set[Id[Organization]]] {
+          case (acc, orgSet) => acc.intersect(orgSet)
+        }.flatMap {
+          case commonOrgs if commonOrgs.nonEmpty => Some(commonOrgs.last) // track an arbitrary common org
```
Comment (ryanpbrewster) [inline kifi-backend/modules/heimdal/app/com/keepit/heimdal/EventContextHelper.scala]: Why not `.head` or `Random.shuffle(commonOrgs).head`?

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/MessagingCommander.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/MessagingCommander.scala:682]: :+1:

File: kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/MessagingCommander.scala (hunk 4 above)
Comment (camhashemi) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/MessagingCommander.scala]: I can do `Random.shuffle`.. this is just what we've chosen to have one `orgId` tracked (afaik we don't/can't send arrays/hashmaps of data to amplitude)

Comment (ryanpbrewster) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/MessagingCommander.scala]: But why `.last` instead of `.head`?

Comment (camhashemi) [inline kifi-backend/modules/eliza/app/com/keepit/eliza/commanders/MessagingCommander.scala]: why not?

**Pull request merged**